scripts/commands/checkLinks.ts

```diff
@@ -0,0 +1,140 @@
+const CONTENT_FILE_EXTENSIONS = [".md", ".mdx", ".ipynb"]
+
+class Link {
+  readonly value: string
+  readonly anchor: string
+  readonly origin: string
+  readonly isExternal: boolean
+
+  constructor(linkString: string, origin: string) {
+    /* 
+     * linkString: Link as it appears in source file
+     *     origin: Path to source file containing link
+     */
+
+    const splitLink = linkString.split('#', 1)
+    this.value = splitLink[0]
+    this.anchor = (splitLink.length > 1) ? `#${splitLink[1]}` : ''
+    this.origin = origin
+    this.isExternal = linkString.startsWith("http")
+  }
+
+  resolve(): string[] {
+    /*
+     * Return list of possible paths link could resolve to
+     */
+    if ( this.isExternal ) { return [ this.value ] }
+    if ( this.value === '' ) { return [ this.origin ] }  // link is just anchor
+    if ( this.value.startsWith("/images") ) {
+      return [ path.join("public/", this.value) ]
+    }
+
+    let baseFilePath
+    if (this.value.startsWith('/')) {
+      // Path is relative to DOCS_ROOT
+      baseFilePath = path.join(DOCS_ROOT, this.value)
+    } else {
+      // Path is relative to origin file
+      baseFilePath = path.join(
+        path.dirname(this.origin),
+        this.value
+      )
+    }
+    // Remove trailing '/' from path.join
+    baseFilePath = baseFilePath.replace(/\/$/gm, '')
+
+    // File may have one of many extensions (.md, .ipynb etc.), and/or be
+    // directory with an index file (e.g. `docs/build` should resolve to
+    // `docs/build/index.mdx`). We return a list of possible filenames.
+    let possibleFilePaths = []
+    for (let index of ['', '/index']) {
+      for (let extension of CONTENT_FILE_EXTENSIONS) {
+        possibleFilePaths.push(
+            baseFilePath + index + extension
+        )
+      }
+    }
+    return possibleFilePaths
+  }
+
+  check(filePathCache: string[] = []): boolean {
+    /*
+     * True if link points to existing file, otherwise false
+     * filePathCache: array of known existing files (to reduce disk I/O)
+     */
+    if (this.isExternal) {
+      // External link checking not supported yet
+      return true
+    }
+
+    const possiblePaths = this.resolve()
+    for (let filePath of possiblePaths) {
+      if (filePathCache.includes(filePath)) {
+        return true
+      }
+    }
+    // Check disk for files not in cache (images etc.)
+    for (let filePath of possiblePaths) {
+      if (existsSync(filePath)) {
+        return true
+      }
+    }
+
+    console.log(`❌ ${this.origin}: Could not find link '${this.value}'`)
+    return false
+  }
+}
+
+function markdownFromNotebook(source: string): string {
+  let markdown = ''
+  for (let cell of JSON.parse(source).cells) {
+    if (cell.source === 'markdown') {
+      markdown += cell.source
+    }
+  }
+  return markdown
+}
+
+function checkLinksInFile(filePath: string, filePaths: string[]): boolean {
+  const source = readFileSync(filePath, {encoding: 'utf8'})
```

Comment:
We'll still want this to be an async read, which will require marking `async function checkLinksInFile`. This line will become `const source = await readFile(...)`

Then, lines 130-133 will become:

```typescript
const results = await Promise.all(filePaths.map(fp => checkLinksInFile(fp, filePaths))
if (results.some(x => !x)) {
}
```

That will parallelize things.